### ISSUE TITLE
[Enhancement]: Improve the compaction selection policy

### ISSUE THREAD

**Issue by xiaofan-luan** [milvus-io/milvus#23588]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What would you like to be added?

We've seen a lot of issues on compaction selection policy, and those issues causing major problems in:
1. Frequent deletes/ Updates.
2. Performance degrade due to segment number is huge.
3. Too frequent compactions when there are many small flushes. 

So I guess we need to think of something we can do, to reduce compaction pressure for small datanodes and increase segment size if the data becomes stable.

**Comment by xiaofan-luan** [comment]
1. dataCoord.segment.smallProportion is 0.5, which means any segment larger than half of the segment will not be able to merged into a large segment, I thought it should change to 0.8 or something.
2. tune min segment to merge based on how many compaction tasks in flight right now, if there are many small segments then we will set large MinSegmentToMerge. If the cluster is free then we will try to merge 2 segments 
3. Concurrently run the compactions, but be careful of each segment is involved in two compactions and it will cause segment not found.

**Comment by xiaofan-luan** [comment]
The overall goal is to make sure after compaction, each of the segment should be close to the segment size limit (70%-90%) fullness.

**Comment by xiaofan-luan** [comment]
We can probably relax the limitation of segment size a little bit also.
Say if two segments are gonna to merge to a 550MB, we should just do it. 

I thought we can relax the limitation by 1.1X so largest segment won't pass 512 * 1.1 ~ 560MB.

**Comment by xiaofan-luan** [comment]
> We can probably relax the limitation of segment size a little bit also. Say if two segments are gonna to merge to a 550MB, we should just do it.
> 
> I thought we can relax the limitation by 1.1X so largest segment won't pass 512 * 1.1 ~ 560MB.

Also we should stop using row count but turned into segment size.

**Comment by XuanYang-cn** [comment]
After #23692 , I'll invite qa-team to help verify the results.

**Comment by XuanYang-cn** [comment]
> We can probably relax the limitation of segment size a little bit also. Say if two segments are gonna to merge to a 550MB, we should just do it.
> 
> I thought we can relax the limitation by 1.1X so largest segment won't pass 512 * 1.1 ~ 560MB.

@xiaofan-luan It's 1.25X by default, do you think its too big?

**Comment by xiaofan-luan** [comment]
keep it open

**Comment by xiaofan-luan** [comment]
keep it

**Comment by xiaofan-luan** [comment]
keep it

**Comment by xiaofan-luan** [comment]
keep it open

**Comment by xiaofan-luan** [comment]
keep it open

**Comment by xiaofan-luan** [comment]
keep it
